New properties and attributes for toric line bundles

HereAround opened the pull request:
- Activate caching for `ToricLineBundles,`
- Add divisor of `ToricLineBundle,`
- Add degree of `ToricLineBundle,`
- Derive several properties of `ToricLineBundles` from its (a) divisor.

This PR contains the changes of https://github.com/oscar-system/Oscar.jl/pull/940 and https://github.com/oscar-system/Oscar.jl/pull/929 to simplify a possible rebase.

HereAround committed:
Allow caching for ToricLineBundles

HereAround committed:
Add divisor of ToricLineBundle

HereAround committed:
Derive several properties of ToricLineBundles from its/a divisor

fingolfin commented on src/ToricVarieties/LineBundles/properties.jl:
```diff
@@ -0,0 +1,87 @@
+"""
+function isbasepoint_free(l::ToricLineBundle)
+    return get_attribute!(l, :isbasepoint_free) do
+        return isbasepoint_free(divisor(l))
```
If the divisor caches this property (does it?), then do we really need to cache it here? I don't think it'll be substantially faster (though we could measure).

This would be different if we had "real" properties as in GAP, where (a) storing a property only takes 2 bits (one for its value, and one to know whether the value was set), and (b) there was a real "implication system" as in GAP. Perhaps we'll want to add such a thing at some point, but for now... ?

HereAround replied:
Good point. I will adjust (and find out why all tests fail...)

HereAround committed:
Add method degree for ToricLineBundle

HereAround commented:
This is ready for review now.

fingolfin commented on src/ToricVarieties/LineBundles/attributes.jl:
```diff
@@ -21,3 +22,36 @@ function variety(l::ToricLineBundle)
+        map1 = map_from_cartier_divisor_group_to_picard_group(variety(l))
+        map2 = map = map_from_cartier_divisor_group_to_torus_invariant_divisor_group(variety(l))
+        image = map2(preimage(map1, class)).coeff
+        coeffs = vec([Int(x) for x in image])
```
So `image` is an `fmpz_mat` row vector here. Is it really guaranteed that all entries will fit into an `Int?

HereAround replied:
Perse not. However, I believe we are currently limited to cases for which this cast (hopefully) works. Namely, the list of coefficients is eventually passed to `Polymake`:

> ptd = Polymake.fulton.TDivisor(COEFFICIENTS=coeffs)

If I call this with `coeffs = [fmpz(1), fmpz(2)]`, then this fails with

> ERROR: ArgumentError: Unrecognized argument type: fmpz.

Instead, with `coeffs = [1, 2]` this works.

thofma replied:
What happens if you do `BigInt(x)`?

HereAround replied:
`coeffs = [BigInt(1), BigInt(2)]` seems to work as input for 

> ptd = Polymake.fulton.TDivisor(COEFFICIENTS=coeffs)

thofma replied:
Not sure if it is realistic that an invariant divisor is bigger than `2^63`, but with `BigInt` we would be on the "safe side". Not sure if it has performance implications though.

fingolfin replied:
If we stay with Int, we need an overflow check. But I really think we should store fmpz, and ToricDivior should accept these; and then either Polymake needs to learn to accept these here (cc @benlorenz ) or we convert to BigInt. (If the coeffs are not accessed externally later, we can of course immediately work with BigInt here, whatever

HereAround replied:
I think I would prefer fmpz.

(Similar issue: rays of a fan currently returns `Polymake.rational`, for which I would prefer fmpz, see https://github.com/oscar-system/Oscar.jl/pull/939 and the issue encountered by @fieker in https://github.com/oscar-system/Oscar.jl/issues/932).

HereAround replied:
The coefficients can be accessed externally (for all it matters).

fingolfin replied:
Agreed

benlorenz replied:
`Polymake.fulton.TDivisor` is a direct interface to a pure `polymake` construction (from autogenerated code) which cannot deal with fmpz types (yet?). <del>Until there is some kind of Oscar wrapper around the TDivisor / or</del> until we can inject some kind of automatic conversion there, this needs to be passed polymake compatible types as input, i.e. Int64, BigInt, Polymake.Rational, Polymake.Integer.
Inside polymake this will be stored as a polymake vector of GMP rationals anyway, so just cast them to some big enough type:
```julia
julia> ptd = Polymake.fulton.TDivisor(COEFFICIENTS=Polymake.Vector{Polymake.Integer}([fmpz(1), fmpz(2)]))
type: TDivisor

COEFFICIENTS
        1 2

julia> ptd.COEFFICIENTS
pm::Vector<pm::Rational>
1 2
```

fingolfin replied:
Thanks for the explanation, Benjamin

fingolfin commented on src/ToricVarieties/LineBundles/properties.jl:
```diff
@@ -0,0 +1,71 @@
+@doc Markdown.doc"""
+    istrivial(l::ToricLineBundle)
+
+Returns if the toric line bundle `l` is trivial.
```
Meta remark (applies to other docstrings, and so I don't expect you to change this here): this should be reformulated as `Returns a boolean indicating wether...` or `Returns true if ...`  -- right now it sounds as if for non-trivial bundles, the function might not return at all ;-)

HereAround replied:
Thank you. Good point. I will adjust this in a future PR for the other docstrings.

HereAround replied:
I just modified these documentations for the new properties.

thofma commented on src/ToricVarieties/LineBundles/constructors.jl:
```diff
@@ -23,7 +23,7 @@ Construct the line bundle on the abstract normal toric variety `v` with class `c
 function ToricLineBundle(v::AbstractNormalToricVariety, input_class::Vector{fmpz})
     class = picard_group(v)(input_class)
-    return ToricLineBundle(v, class)
+    return ToricLineBundle(v, class, Dict())
```
I think it would be easier to add a default 2-argument constructor, see the "note" in https://nemocas.github.io/AbstractAlgebra.jl/dev/misc/#AbstractAlgebra.@attributes.

HereAround replied:
I am confused. The above is a 2-argument constructor. What is wrong with that?

thofma replied:
Sorry, I mean that if you add a default inner constructor in line 9, then you can just do `ToricLineBundle(v, class)` here. See the example in the note.

thofma commented:
I think it would make sense to keep `coefficients` instead of `coeffs`, since we now use this everywhere else (hopefully). There are lots of `.coeffs` fields flying around, but these are internal and can be as inconsistent as they want :)

HereAround commented:
> I think it would make sense to keep `coefficients` instead of `coeffs`, since we now use this everywhere else (hopefully). There are lots of `.coeffs` fields flying around, but these are internal and can be as inconsistent as they want :)

That is ok. This change can be revoked. No problem.

thofma commented on src/ToricVarieties/LineBundles/constructors.jl:
```diff
@@ -4,7 +4,7 @@
-struct ToricLineBundle <: ToricCoherentSheaf
+@attributes mutable struct ToricLineBundle <: ToricCoherentSheaf
     variety::AbstractNormalToricVariety
     divisor_class::GrpAbFinGenElem
```
```suggestion
    divisor_class::GrpAbFinGenElem
    
    ToricLineBundle(variety::AbstractNormalToricVariety, divisor_class::GrpAbFinGenElem) = new(variety, divisor_class)
```
Then you don't need the `Dict()` below.

HereAround replied:
I think I see this now. Thank you for elaborating. I will try this out later and then update this PR.

fingolfin commented on src/ToricVarieties/LineBundles/attributes.jl:
```diff
@@ -21,3 +22,36 @@ function variety(l::ToricLineBundle)
+"""
+function degree(l::ToricLineBundle)
+    return get_attribute!(l, :degree) do
+        return sum(coeffs(toric_divisor(l)))::Int
```
I guess this also should be `fmpz`?

fingolfin replied:
Actually since Julia should be able to infer that `coeffs(toric_divisor(l))` is a `Vector{Int}` (and perhaps soon a `Vector{fmpz}` ?) it should also infer the return type here correctly.

HereAround committed:
Use fmpz for ToricDivisors

HereAround committed:
Add constructors in structs of toric line bundles, divisors and varieties
-> Remove then obsolete dicts

HereAround committed:
Enforce name convention: toricvariety(D) -> toric_variety(D) for toric divsior D

HereAround commented:
I have just rebased this PR and tried to address all of the points raised above:
1. https://github.com/oscar-system/Oscar.jl/pull/941/commits/b5b3bdc4d6d926b5b003a92ede5e62470ad8a503 ensures that we use fmpz for toric divisors. This uses the above suggestion by @benlorenz.
2. As pointed out by @fingolfin , also the return value of the degree of a line bundle has been changed from `::Int` to `::fpmz` in see https://github.com/oscar-system/Oscar.jl/pull/941/commits/3a8fd5a1d2a1b75812f5137c301dace9e153fa0d.
3. @thofma: As you suggested, I added constructors to the `strucs` and removed the then obsolete `Dict()`. This applied toric-variety-wide and has been changed in https://github.com/oscar-system/Oscar.jl/pull/941/commits/80a78dfeae522a3d87b57d3879bc3663e6c41079.

Once the tests go green (which I hope should happen shortly), this PR is again ready for review.

fingolfin commented on src/ToricVarieties/LineBundles/properties.jl:
```diff
@@ -0,0 +1,73 @@
+
+
+@doc Markdown.doc"""
+    isbasepoint_free(l::ToricLineBundle)
```
```suggestion
    is_basepoint_free(l::ToricLineBundle)
```

HereAround replied:
Thank you @fingolfin. Let me adjust this... Done.

fingolfin reviewed (approved):
Seems OK to me. One tiny typo
